### PR TITLE
[openvpn] Distroless

### PR DESCRIPTION
**Pull request by vitaliy-sn**
## Description

Images are now based on a distroless image.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: openvpn
type: feature
summary: Images are based on a distroless image.
```

### CODE REVIEW

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = netLinkCreateTuntap(fmt.Sprintf("tun-%s", protocol), 1400)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	routeAdd(network, fmt.Sprintf("tun-%s", protocol), routeTable)
```
Comment (name212) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:141]: please move  `fmt.Sprintf("tun-%s", protocol)` to variable and use this variable also in beyond

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	args = append(args, "--config")
+	args = append(args, "/etc/openvpn/openvpn.conf")
+	args = append(args, "--proto")
+	args = append(args, protocol)
+	args = append(args, "--management")
+	args = append(args, "127.0.0.1")
+	args = append(args, mgmtport)
+	args = append(args, "--dev")
+	args = append(args, fmt.Sprintf("tun-%s", protocol))
+	log.Println(args)
```
Comment (name212) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:165]:
```suggestion
	log.Println("Openvpn args", args)
```

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+	}
+
+	command := fmt.Sprintf("/bin/mknod")
+	var args []string
+	args = append(args, "/dev/net/tun")
+	args = append(args, "c")
+	args = append(args, "10")
+	args = append(args, "200")
+	cmd := exec.Command(command, args...)
+	_, err = cmd.CombinedOutput()
```
Comment (name212) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:224]: we do not need output here. Please use `cmd.Run` instead `CombinedOutput`

File: modules/500-openvpn/images/ovpn-admin/werf.inc.yaml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  install:
+    - mkdir -p /src
+    - cd /src
+    - git clone https://github.com/flant/ovpn-admin.git .
+    - git checkout {{ $gitCommit }}
+    - find /patches -name '*.patch' -exec git apply {} \;
+    - git status
+    - git diff
+    - echo {{ $gitCommit }} > version
+    - echo "rebuild 123"
```
Comment (name212) [inline modules/500-openvpn/images/ovpn-admin/werf.inc.yaml:34]: Is it necessary?

File: modules/500-openvpn/images/ovpn-admin/werf.inc.yaml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    - export GO_VERSION=${GOLANG_VERSION}
```
Comment (name212) [inline modules/500-openvpn/images/ovpn-admin/werf.inc.yaml:35]: i think it is unnecessary because go version contains in image

**Review by name212 — CHANGES_REQUESTED** [review]
some nits

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+	switch protocol {
+	case "tcp":
+		mgmtport = "8989"
+		routeTable = 10
+	case "udp":
+		mgmtport = "9090"
+		routeTable = 11
+	default:
+		log.Fatalf("OPENVPN_PROTO env value must be tcp or udp: %s", protocol)
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:64]:
```suggestion
		log.Fatalf("OPENVPN_PROTO env value must be set to 'tcp' or 'udp': %s", protocol)
```

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+	var args []string
+	args = append(args, "--config")
+	args = append(args, "/etc/openvpn/openvpn.conf")
+	args = append(args, "--proto")
+	args = append(args, protocol)
+	args = append(args, "--management")
+	args = append(args, "127.0.0.1")
+	args = append(args, mgmtport)
+	args = append(args, "--dev")
+	args = append(args, fmt.Sprintf("tun-%s", protocol))
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:164]:
```suggestion
	args := []string{
	   "--config",
	   "/etc/openvpn/openvpn.conf",
	   "--proto",
	   protocol,
	   "--management",
	   "127.0.0.1",
	   mgmtport,
	   "--dev",
	   "tun-" + protocol,
     }
```

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+	}
+	if !ok {
+		err := iptablesMgr.Insert(table, chain, pos, rule...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:199]:
```suggestion
	if ok {
	  return nil
	}
	
  return iptablesMgr.Insert(table, chain, pos, rule...)
}
```

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+		}
+	}
+
+	command := fmt.Sprintf("/bin/mknod")
+	var args []string
+	args = append(args, "/dev/net/tun")
+	args = append(args, "c")
+	args = append(args, "10")
+	args = append(args, "200")
+	cmd := exec.Command(command, args...)
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:223]: Same case as above

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+	if err != nil {
+		return err
+	}
+
+	link, _ := netlink.LinkByName(linkAttrs.Name)
+	err = netlink.LinkSetUp(link)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:251]:
```suggestion
	return netlink.LinkSetUp(link)
```

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+func routeAdd(dstNet string, linkName string, table int) {
+	dstIPNet, err := parseIPNet(dstNet)
+	if err != nil {
+		log.Fatal("error parse IPNet: ", err)
+	}
+
+	link, _ := netlink.LinkByName(linkName)
+	if err != nil {
+		log.Fatal("error parse IPNet: ", err)
+	}
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:263]: duplicate with string 256

File: modules/500-openvpn/images/openvpn/werf.inc.yaml [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    - GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -w" -o /entrypoint entrypoint.go
+---
+artifact: {{ .ModuleName }}/openssl-artifact
+from: {{ .Images.BASE_ALPINE }}
+shell:
+  install:
+    - apk add --no-cache curl git autoconf automake make libtool g++ linux-headers libnl3-dev libcap-ng-dev lz4-dev linux-pam-dev tree
+    - mkdir -p /src
+    - cd /src
+    - git clone --depth 1 -b openssl-3.1.3 https://github.com/openssl/openssl.git
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/werf.inc.yaml:35]: We should use fox repo for this

File: modules/500-openvpn/images/openvpn/werf.inc.yaml [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    - mkdir -p /src
+    - cd /src
+    - mkdir lzo
+    - cd lzo
+    - curl -sL https://www.oberhumer.com/opensource/lzo/download/lzo-2.10.tar.gz | tar -xvz --strip-components=1
+    - ./configure --enable-static --disable-debug
+    - make -j4
+    - make install
+    - cd /src
+    - git clone --depth 1 -b v2.5.6 https://github.com/OpenVPN/openvpn.git
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/werf.inc.yaml:60]: we should use fox repo for this and use {{.SOURCE_REPO}} var

File: modules/500-openvpn/images/openvpn/werf.inc.yaml [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  to: /
+  before: install
+shell:
+  install:
+    - apk add --no-cache curl git autoconf automake make libtool g++ linux-headers libnl3-dev libcap-ng-dev lz4-dev linux-pam-dev tree
+    - mkdir -p /src
+    - cd /src
+    - mkdir lzo
+    - cd lzo
+    - curl -sL https://www.oberhumer.com/opensource/lzo/download/lzo-2.10.tar.gz | tar -xvz --strip-components=1
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/openvpn/werf.inc.yaml:55]: should be in the fox repo

File: modules/500-openvpn/images/ovpn-admin/werf.inc.yaml [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+mount:
+- fromPath: ~/go-pkg-cache
+  to: /go/pkg
+shell:
+  beforeInstall:
+    - apk add --no-cache git
+  install:
+    - mkdir -p /src
+    - cd /src
+    - git clone https://github.com/flant/ovpn-admin.git .
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/ovpn-admin/werf.inc.yaml:28]: should be in the fox repo

File: modules/500-openvpn/images/ovpn-admin/werf.inc.yaml [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+  stageDependencies:
+    install:
+      - '**/*'
+shell:
+  beforeInstall:
+    - apk add --no-cache git
+  install:
+    - mkdir -p /src
+    - cd /src
+    - git clone https://github.com/flant/ovpn-admin.git .
```
Comment (RomanenkoDenys) [inline modules/500-openvpn/images/ovpn-admin/werf.inc.yaml:55]: fox repo

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some nits

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+		Rule:  strings.Fields("-j CONNMARK --restore-mark"),
+		Pos:   1,
+	}
+
+	err = insertUnique(iptablesMgr, nat.Table, nat.Chain, nat.Rule, nat.Pos)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = insertUnique(iptablesMgr, manglePreroutingSetMark.Table, manglePreroutingSetMark.Chain, manglePreroutingSetMark.Rule, manglePreroutingSetMark.Pos)
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 1 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 2 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 3 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/ovpn-admin/werf.inc.yaml (hunk 4 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/ovpn-admin/werf.inc.yaml]: It's necessary.
>- echo {{ $gitCommit }} > version

The rest has been removed.

File: modules/500-openvpn/images/ovpn-admin/werf.inc.yaml (hunk 5 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/ovpn-admin/werf.inc.yaml:32]: We use these environment variables in all distroless builds

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 6 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 7 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 8 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 10 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 11 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/werf.inc.yaml (hunk 12 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/werf.inc.yaml]: done

File: modules/500-openvpn/images/openvpn/werf.inc.yaml (hunk 13 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/werf.inc.yaml]: done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Comment by vitaliy-sn** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 7538637677 7538637677-1-con-1-25 /sys/deckhouse-oss/install:pr6245 centos@3.73.97.27

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've recognized your '/e2e/abort' command and started the workflow...

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
unresolved questions

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+	rule := netlink.NewRule()
+	rule.Table = routeTable
+	rule.Priority = 1
+	rule.Mark = routeTable
+	err = netlink.RuleAdd(rule)
+
+	if err != nil {
+		if !os.IsExist(err) {
+			log.Fatal(err)
+		}
```
Comment (aleksey-su) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:130]: Maybe add error logging to else?

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+	return netlink.LinkSetUp(link)
+}
+
+func routeAdd(dstNet string, linkName string, table int) {
+	dstIPNet, err := parseIPNet(dstNet)
+	if err != nil {
+		log.Fatal("error parse IPNet: ", err)
+	}
+
+	link, _ := netlink.LinkByName(linkName)
```
Comment (aleksey-su) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:259]:
```suggestion
	link, err := netlink.LinkByName(linkName)
```

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+
+func waitingForFile(path string) {
+	for {
+		_, err := os.Stat(path)
+		if errors.Is(err, os.ErrNotExist) {
+			log.Println("waiting for", path)
+			time.Sleep(2 * time.Second)
+		} else {
+			break
+		}
```
Comment (aleksey-su) [inline modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:290]:
```suggestion
		_, err := os.Stat(path)
		if errors.Is(err, os.ErrNotExist) {
			log.Println("waiting for", path)
			time.Sleep(2 * time.Second)
			continue
		} 
		break
```

**Review by aleksey-su — CHANGES_REQUESTED** [review]

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 18 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go:130]: Everything is logged here, except when the rule exists. If the rule exists, then in my opinion this is not an error.

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 19 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

File: modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go (hunk 20 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go]: done

**Review by aleksey-su — APPROVED** [review]

**Review by RomanenkoDenys — APPROVED** [review]

**Review by name212 — APPROVED** [review]

**Pull request merged**